lib/model/model.go

```diff
@@ -850,78 +850,155 @@ func (m *Model) ClusterConfig(deviceID protocol.DeviceID, cm protocol.ClusterCon
 		}
 	}
 
-	var changed bool
+	if deviceCfg := m.cfg.Devices()[deviceID]; deviceCfg.Introducer {
+		m.handleIntroducer(deviceCfg, cm)
+	}
+}
 
-	if m.cfg.Devices()[deviceID].Introducer {
-		// This device is an introducer. Go through the announced lists of folders
-		// and devices and add what we are missing.
+// handleIntroducer handles adding and removing devices/shares that are shared
+// by an introducer device
+func (m *Model) handleIntroducer(cfg config.DeviceConfiguration, cm protocol.ClusterConfig) {
+	// This device is an introducer. Go through the announced lists of folders
+	// and devices and add what we are missing, remove what we have extra.
+	changed := false
+	foldersDevices := make(map[string][]protocol.Device)
```

Comment:
I suggest the following utility type to avoid all the nested loops and byte compares below:

```go

// folderDeviceSet is a set of (folder, deviceID) pairs
type folderDeviceSet map[string]map[protocol.DeviceID]struct{}

// set adds the (dev, folder) pair to the set
func (s folderDeviceSet) set(dev protocol.DeviceID, folder string) {
	devs, ok := s[folder]
	if !ok {
		devs = make(map[protocol.DeviceID]struct{})
		s[folder] = devs
	}
	devs[dev] = struct{}{}
}

// has returns true if the (dev, folder) pair is in the set
func (s folderDeviceSet) has(dev protocol.DeviceID, folder string) bool {
	_, ok := s[folder][dev]
	return ok
}

// has returns true if the (dev, folder) pair is in the set, where dev is a
// raw byte representation of the device ID
func (s folderDeviceSet) hasRaw(dev []byte, folder string) bool {
	if len(dev) != 32 {
		panic("not a device id")
	}
	var devid protocol.DeviceID
	copy(devid[:], dev)
	_, ok := s[folder][devid]
	return ok
}
```

It could also be used as `m.folderDevices` as it's still rangeable and stuff.